File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/network/http/CountingOutputStream.java
package io.embrace.android.embracesdk.internal.network.http;

import java.io.ByteArrayOutputStream;
import java.io.FilterOutputStream;
import java.io.IOException;
import java.io.OutputStream;

/**
 * Counts the number of bytes written to the output stream, also captures the request body.
 */
class CountingOutputStream extends FilterOutputStream {
    private long count;

    ByteArrayOutputStream os = new ByteArrayOutputStream();

    /**
     * Wraps another output stream, counting the number of bytes written.
     *
     * @param out the output stream to be wrapped
     */
    public CountingOutputStream(OutputStream out) {
        super(out);
    }

    /**
     * Returns the number of bytes written.
     */
    public long getCount() {
        return count;
    }

    /**
     * Returns the request body written.
     */
    byte[] getRequestBody() {
        return os.toByteArray();
    }

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
        out.write(b, off, len);
        count += len;
        if (b != null) {
            os.write(b, off, len);
        }
    }

    @Override
    public void write(int b) throws IOException {
        out.write(b);
        count++;
        os.write(b);
    }

    // Overriding close() because FilterOutputStream's close() method pre-JDK8 has bad behavior:
    // it silently ignores any exception thrown by flush(). Instead, just close the delegate stream.
    // It should flush itself if necessary.
    @Override
    public void close() throws IOException {
        out.close();
    }
}

File: embrace-android-infra/src/main/kotlin/io/embrace/android/embracesdk/internal/utils/CountingOutputStream.kt
package io.embrace.android.embracesdk.internal.utils

import java.io.OutputStream

/**
 * Counts the bytes written through it
 */
class CountingOutputStream(private val delegate: OutputStream) : OutputStream() {

    var written: Long = 0
        private set

    override fun write(b: Int) {
        delegate.write(b)
        written++
    }

    override fun write(b: ByteArray, off: Int, len: Int) {
        delegate.write(b, off, len)
        written += len
    }

    override fun flush() = delegate.flush()

    override fun close() = delegate.close()
}
